### PR TITLE
Add a cycle detector for generic `Graph`s and `mir::Body`s

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Cycle detection is one way to differentiate the upcoming `const_loop` feature flag (#52000) from the `const_if_match` one (#49146). It would be possible to use the existing implementation of strongly-connected components for this but less efficient.

The ["tri-color" terminology](http://www.cs.cornell.edu/courses/cs2112/2012sp/lectures/lec24/lec24-12sp.html) is common in introductory data structures and algorithms courses: black nodes are settled, grey nodes are visited, and white nodes have no state. This particular implementation is iterative and uses a well-known technique where "node settled" events are kept on the stack alongside nodes to visit. When a settled event is popped, we know that all successors of that node have been visited and themselves settled. If we encounter a successor node that has been visited (is on the stack) but not yet settled, we have found a cycle.

r? @eddyb

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
This probably doesn't need to be reviewed by @eddyb (I r?-ed out of habit).

**Comment by ecstatic-morse** [comment]
Maybe r? @Centril

File: src/librustc_data_structures/graph/iterate/mod.rs [hunk 1]
```diff
@@ -85,3 +85,150 @@ where
...
 }
+
+#[derive(Clone, Copy, Debug)]
+pub enum ControlFlow<T> {
+    Break(T),
+    Continue,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum NodeState {
```
Comment (Centril) [inline src/librustc_data_structures/graph/iterate/mod.rs]: Adding some docs to the type also could be good.

**Comment by Centril** [comment]
> Maybe r? @Centril

Wouldn't even know where to begin. =P

r? @pnkfelix

**Comment by oli-obk** [comment]
r? @oli-obk 

Some more docs would indeed be great.

**Comment by ecstatic-morse** [comment]
I added some more docs comparing this depth-first search to the one in [Introduction to Algorithms](https://en.wikipedia.org/wiki/Introduction_to_Algorithms). Hopefully this is more understandable.

**Comment by oli-obk** [comment]
@bors r+

Thanks, that helps

**Pull request merged**